### PR TITLE
SI-4842 Forbid access to in-construction this in self-constructor args

### PR DESCRIPTION
**Pull request by retronym**
The check was already in place for direct calls to the super constructor.
Without this additional check, ExplicitOuter crashes, as it doesn't create
an $outer pointer for the constructor-arg scoped inner object, but expects
one to exist when transforming the Outer.this reference.

review by @hubertp
- (how) should I report more than one such error per file?

### CODE REVIEW

**Comment by hubertp** [comment]
Could you be more specific? You use pending which is ok, also the positions for the messages are determined by tree.pos, so if you have trees with the same message but different positions these are reported correctly.

Other than that, LGTM.

**Comment by retronym** [comment]
When I tried to collapse these `t4842a.scala` and `t4842b.scala` into a single test, only the first error was reported.

**Comment by hubertp** [comment]
Ah, ok. In that case something fishy is going on. I will have a look but I am on holiday atm.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -1884,7 +1881,39 @@ trait Typers extends Modes with Adaptations with Tags {
       pending.foreach(ErrorUtils.issueTypeError)
     }
 
-    /** Check if a structurally defined method violates implementation restrictions.
+    // Check for SI-4842.
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1884]: it'd be nice to expand the description for that issue here

**Pull request merged**